core/src/layers/prometheus.rs

```diff
@@ -98,784 +95,235 @@ use crate::*;
 
     fn layer(&self, inner: A) -> Self::LayeredAccess {
-        let meta = inner.info();
-        let scheme = meta.scheme();
-
-        PrometheusAccessor {
-            inner,
-            stats: Arc::new(PrometheusMetrics::new(
-                self.registry.clone(),
-                self.requests_duration_seconds_buckets.clone(),
-                self.bytes_total_buckets.clone(),
-                self.path_label_level,
-            )),
-            scheme,
-        }
+        let interceptor = PrometheusInterceptor::register(
+            self.registry.clone(),
+            self.operation_duration_seconds_buckets.clone(),
+            self.operation_bytes_buckets.clone(),
+            self.path_label_level,
+        );
+        observe::MetricsLayer::new(interceptor).layer(inner)
     }
 }
 
-/// [`PrometheusMetrics`] provide the performance and IO metrics.
-#[derive(Debug)]
-pub struct PrometheusMetrics {
-    /// Total times of the specific operation be called.
-    pub requests_total: GenericCounterVec<AtomicU64>,
-    /// Latency of the specific operation be called.
-    pub requests_duration_seconds: HistogramVec,
-    /// Size of the specific metrics.
-    pub bytes_total: HistogramVec,
-    /// The Path Level we will keep in the path label.
-    pub path_label_level: usize,
+#[derive(Clone, Debug)]
+pub struct PrometheusInterceptor {
+    operation_duration_seconds: HistogramVec,
+    operation_bytes: HistogramVec,
+    operation_errors_total: GenericCounterVec<AtomicU64>,
+    path_label_level: usize,
 }
 
-impl PrometheusMetrics {
-    /// new with prometheus register.
-    pub fn new(
+impl PrometheusInterceptor {
+    fn register(
         registry: Registry,
-        requests_duration_seconds_buckets: Vec<f64>,
-        bytes_total_buckets: Vec<f64>,
+        operation_duration_seconds_buckets: Vec<f64>,
+        operation_bytes_buckets: Vec<f64>,
         path_label_level: usize,
     ) -> Self {
-        let labels = if path_label_level > 0 {
-            vec!["scheme", "operation", "path"]
+        let fixed_labels = vec![
+            observe::LABEL_SCHEME,
+            observe::LABEL_NAMESPACE,
+            observe::LABEL_ROOT,
+            observe::LABEL_OPERATION,
+        ];
+        let (labels, labels_with_error) = if path_label_level > 0 {
+            (
+                {
+                    let mut labels = fixed_labels.clone();
+                    labels.push(observe::LABEL_PATH);
+                    labels
+                },
+                {
+                    let mut labels = fixed_labels.clone();
+                    labels.push(observe::LABEL_ERROR);
+                    labels.push(observe::LABEL_PATH);
+                    labels
+                },
+            )
         } else {
-            vec!["scheme", "operation"]
+            (fixed_labels.clone(), {
+                let mut labels = fixed_labels;
+                labels.push(observe::LABEL_ERROR);
+                labels
+            })
```

Comment:
How about implementing in this way:

```rust
// Use maximum labels here to avoid extra alloc.
let mut labels = Vec::with_capacity(6);
vec.extend([
    observe::LABEL_SCHEME,
    observe::LABEL_NAMESPACE,
    observe::LABEL_ROOT,
    observe::LABEL_OPERATION]);
if path_label_level > 0 {
   labels.push(observe::LABEL_PATH);
}

let operation_duration_seconds = register_histogram_vec_with_registry!(&labels);
let operation_bytes = register_histogram_vec_with_registry!(&labels);
...

labels.push(observe::LABEL_ERROR);
let operation_errors_total = register_int_counter_vec_with_registry!(&abels);
```


This way, we can improve code readability and eliminate unnecessary allocation or cloning.